Mage.Sets/src/mage/cards/a/ArenaOfGlory.java

```diff
@@ -0,0 +1,181 @@
+import mage.abilities.condition.common.PermanentsOnTheBattlefieldCondition;
+import mage.abilities.costs.common.ExertSourceCost;
+import mage.abilities.costs.common.TapSourceCost;
+import mage.abilities.costs.mana.ManaCostsImpl;
+import mage.abilities.decorator.ConditionalOneShotEffect;
+import mage.abilities.effects.ContinuousEffectImpl;
+import mage.abilities.effects.OneShotEffect;
+import mage.abilities.effects.common.CreateDelayedTriggeredAbilityEffect;
+import mage.abilities.effects.common.TapSourceEffect;
+import mage.abilities.effects.mana.BasicManaEffect;
+import mage.abilities.keyword.HasteAbility;
+import mage.abilities.mana.RedManaAbility;
+import mage.abilities.mana.SimpleManaAbility;
+import mage.cards.Card;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.*;
+import mage.filter.FilterPermanent;
+import mage.filter.StaticFilters;
+import mage.filter.common.FilterControlledPermanent;
+import mage.game.Game;
+import mage.game.events.GameEvent;
+import mage.game.permanent.Permanent;
+import mage.game.stack.Spell;
+import mage.target.targetpointer.FixedTarget;
+
+import java.util.UUID;
+
+/**
+ * @author Susucr
+ */
+public final class ArenaOfGlory extends CardImpl {
+
+    private static final FilterPermanent filter = new FilterControlledPermanent(SubType.MOUNTAIN);
+    private static final Condition condition
+            = new PermanentsOnTheBattlefieldCondition(filter, ComparisonType.EQUAL_TO, 0);
+
+    public ArenaOfGlory(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.LAND}, "");
+
+        // Arena of Glory enters the battlefield tapped unless you control a Mountain.
+        this.addAbility(new EntersBattlefieldAbility(new ConditionalOneShotEffect(
+                new TapSourceEffect(), condition
+        ), "tapped unless you control a Mountain"));
+
+        // {T}: Add {R}.
+        this.addAbility(new RedManaAbility());
+
+        // {R}, {T}, Exert Arena of Glory: Add {R}{R}. If that mana is spent on a creature spell, it gains haste until end of turn.
+        SimpleManaAbility ability = new SimpleManaAbility(
+                new BasicManaEffect(Mana.RedMana(2)),
+                new ManaCostsImpl<>("{R}")
+        );
+        ability.addEffect(new CreateDelayedTriggeredAbilityEffect(
+                new ArenaOfGloryDelayedTriggeredAbility()
+        ));
+        ability.addCost(new TapSourceCost());
+        ability.addCost(new ExertSourceCost());
+        this.addAbility(ability);
+    }
+
+    private ArenaOfGlory(final ArenaOfGlory card) {
+        super(card);
+    }
+
+    @Override
+    public ArenaOfGlory copy() {
+        return new ArenaOfGlory(this);
+    }
+}
+
+class ArenaOfGloryDelayedTriggeredAbility extends ManaSpentDelayedTriggeredAbility {
+
+    ArenaOfGloryDelayedTriggeredAbility() {
+        super(new ArenaOfGloryTargetEffect(), StaticFilters.FILTER_SPELL_CREATURE);
+        this.usesStack = false;
```

Comment:
That's buggy usage and must be reworked to use replacement style effect, see problems of non-stack delayed triggers here #12111. Standard code from other cards:
* `DelayedTriggeredManaAbility` for additional mana effects on mana usage (resolves immediately, not your use case);
* `ManaSpentDelayedTriggeredAbility` for additional non-mana effects on mana usage (go to stack and resolve later, not your use case);
* Replacement effect on mana usage (watcher or other implementations, your use case - maybe all that cards can have shared implementation as ssk97 mentions above):
![shot_240606_061636](https://github.com/magefree/mage/assets/8344157/36db6ffa-6d8d-4d18-a58b-fa8e4f1abacb)

Reply:
Well, after original topic and problem review -- it can be used as non-stack trigger for some use cases like gain abilities (it will broke some rare combos but it's ok).

Code can be shared as `ManaSpentAndGainAbilityDelayedTriggeredAbility` or like that -- but make sure it will not generate too much instances on multiple mana usage or playable calculation (run game, play with it multiple times and check game state -- is it contains dozen of new continues effects or discarded). Maybe `GainSuspendEffect` can be useful as a reference implementation.

Reply:
For info, the inherited `isInactive` is doing cleanup on player leave or mana pool empty (as there might be mana being kept between turn/phases):
![image](https://github.com/magefree/mage/assets/34709007/d4544151-5d98-4c00-b821-63e0c505e563)

It does not matter for Arena as the haste ability is an instance, but there might be a problem for other kinds of triggers if the land is untapped/tapped multiple times. There might be an accumulation of triggers until the mana pool is empty. Could lead to bugs with other effects, but haste-giving is fine. Need to identify there is not already a DelayedTriggeredAbility in continuous effect maybe?

Anyway I'm going to merge this, and cleanup others cards at a later point. (after a slight rework to use `GainAbilityTargetEffect` as it works for spell on stack + permanent the spell becomes)